### ISSUE TITLE
caching does not happen

### ISSUE THREAD

**Issue by Almenon** [gleitz/howdoi#213]
I was looking into bundling howdoi with [arepl](https://github.com/Almenon/AREPL-vscode) and I came across this issue:

## Environment:

OS: windows 10
python: 3.6
howdoi version 1.1.13
HOWDOI_DISABLE_CACHE env variable is not set

## Expected Result:
Cache appears in C:\Users\myUserName\\.cache\howdoi
Duplicate howdoi queries use the cache and return instantly rather than using the network

## Actual result:
no cache file in C:\Users\myUserName\.cache\howdoi
Fiddler records multiple howdoi queries
There is a 3 second hang before results appear

![image](https://user-images.githubusercontent.com/13080965/47061936-988dfe00-d188-11e8-8d45-b8d02cf3646f.png)

## Investigation
The following code works:

```python
import os
import sys
import requests
import requests_cache

XDG_CACHE_DIR = os.environ.get('XDG_CACHE_HOME',
                               os.path.join(os.path.expanduser('~'), '.cache'))
CACHE_DIR = os.path.join(XDG_CACHE_DIR, 'howdoi')
CACHE_FILE = os.path.join(CACHE_DIR, 'cache{0}'.format(
    sys.version_info[0] if sys.version_info[0] == 3 else ''))

def _enable_cache():
    if not os.path.exists(CACHE_DIR):
        os.makedirs(CACHE_DIR)
    requests_cache.install_cache(CACHE_FILE)

_enable_cache()

howdoi_session = requests.session()

r=howdoi_session.get('http://httpbin.org/delay/1')
# cache will appear in cache folder
```

This is very similar to your code - except with one major difference: in your code _enable_cache is called *after* howdoi_session is assigned. So install_cache monkey patches requests.session, but by then it is too late.  The fix is very simple:

```python
def _enable_cache():
    if not os.path.exists(CACHE_DIR):
        os.makedirs(CACHE_DIR)
    #requests_cache.install_cache(CACHE_FILE) # get rid of this line

    # reassign howdoi with cached session
    howdoi_session = requests_cache.CachedSession(CACHE_FILE)
```
There are alternate solutions depending on how you like to structure your code.

To verify that it worked I conducted numerous speed tests*.

...... That's when I made a horrifying discovery.

Caching _does not actually speed up the program!_

In fact, it **slows it down**!

Caching: ~3 seconds
Current program (broken caching): ~2.5 seconds
All references to caching code & library removed: ~2 seconds

Most of the time of the program is just taken by launching python - making the web request is neglible†.

This brings to mind the famous quote by Donald Knuth

> "Premature optimization is the root of all evil"

That being said, I have a good internet connection. I tried slowing it down using fiddler and some other random tool I found online that's probably using my computer to mine cryptocurrency as we speak, but I wasn't able to get either one working. So I tried on my very old laptop, my first actually, which has a connection so slow it's painful. On there.... it was even faster than my desktop. 

![wat-emoji](https://user-images.githubusercontent.com/13080965/47251339-77c3e380-d3e8-11e8-9301-b42e8c971ea6.png)

But I decided to look into it a bit more, cause who needs sleep?

It turns out my solution, while working, is terrible for performance. Not sure why a simple recreation of session takes so long, but it does. This alternate fix performs _much_ better:

```python
# howdoi_session = requests.session() # line 75, get rid of this

# enable the cache if user doesn't want it to be disabled
if os.getenv('HOWDOI_DISABLE_CACHE'):
    howdoi_session = requests.session()
else: 
    if not os.path.exists(CACHE_DIR):
        os.makedirs(CACHE_DIR)
    howdoi_session = requests_cache.CachedSession(CACHE_FILE)
```

Later on you would get rid of the _enable_cache function and in line 319 you would remove the call to _enable_cache.

Doing this increases performance massively - on my laptop I get runs as fast as a half second. 🚀

† So what I said earlier about most of the time being python launching? That was totally wrong. 

* Disclaimer: perfomance runs were ran with simple useage of print(time()-startTime).  They should serve only as very rough estimates that would vary by environment.

## My recomendations:

You have four options.

1. Fix the caching, as shown above. This will slow down the first time useage of howdoi by a noticeable amount, but there is ~~very little~~ significant performance improvements with same future requests. However, I'm not sure how often users would repeat the same query. I'm guessing the majority of the queries would be new, in which case most of the time caching would slow down the user :(

2. Option 1, but turn caching off by default so users can still opt in if they want.

3. Disable caching, but leave the caching code in place. If needed it would be simple to enable caching again.  The users wouldn't even notice that its disabled given that it wasnt working in the first place.

4. Get rid of the caching entirely. A small perfomance improvement due to not needing to load the caching library and the code becomes simpler. Downside is if the project expands and you _do_ want caching you will have to put it back in again.

Sadly I would recomend 4 but it's up to you what you want to do. I certainly wouldn't mind my fix being taken in :)

**Comment by Almenon** [comment]
I tried it - doesn't have any performance gain :/

Which makes sense - getting the pickled version from a file may skip some initialization logic but then you have to deal with file I/O, which is the snail to the CPU's proverbial cheetah.

**Comment by gleitz** [comment]
Fixed by #214
